File: app/shared/failures.py
"""The one place a user-facing failure message is written.

Why the catalog is server-side and not in the UI: the browser is not the only
thing that shows these. A failed assistant turn is persisted as
`message.error_message` and re-rendered on reload, the SSE stream carries it
live, and the orchestrator logs it. Defining the copy once here means those
three never disagree, and changing the wording is one edit.

The UI still owns *behaviour* - whether to offer a sign-in link, whether to
show a retry button - which it decides from `code`, never from message text.
Matching on message text is what this module exists to replace: the frontend
used to `message.includes('401')` and `includes('inference')` against raw
Python exception strings, so any upstream rewording silently downgraded a
precise message to "Something went wrong."

Contract, everywhere a failure crosses a service boundary:

    {"code": "MODEL_UNAVAILABLE", "message": "<user-facing>", "retryable": true}
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Failure:
    code: str
    message: str
    # Whether trying the same request again could plausibly succeed. Drives the
    # retry affordance in the UI; it is not a promise that a retry will work.
    retryable: bool
    http_status: int

    def payload(self, detail: str | None = None) -> dict:
        """The wire shape. `detail` is for operators, never shown to users."""
        body = {"code": self.code, "message": self.message, "retryable": self.retryable}
        if detail:
            body["detail"] = detail[:300]
        return body


def _failure(code: str, message: str, retryable: bool, http_status: int) -> Failure:
    failure = Failure(code=code, message=message, retryable=retryable, http_status=http_status)
    CATALOG[code] = failure
    return failure


CATALOG: dict[str, Failure] = {}

# ── The caller ──────────────────────────────────────────────────────────────

AUTH_REQUIRED = _failure(
    "AUTH_REQUIRED",
    "Your session could not be verified. Please sign in again.",
    retryable=False,
    http_status=401,
)
REQUEST_INVALID = _failure(
    "REQUEST_INVALID",
    "That request could not be understood. Please rephrase and try again.",
    retryable=False,
    http_status=422,
)
CANCELLED = _failure(
    "CANCELLED",
    "The response was stopped.",
    retryable=True,
    http_status=499,
)

# ── Services in the chain ───────────────────────────────────────────────────

AGENT_UNAVAILABLE = _failure(
    "AGENT_UNAVAILABLE",
    "The assistant is not reachable right now. Please try again in a moment.",
    retryable=True,
    http_status=503,
)
DOMAIN_UNAVAILABLE = _failure(
    "DOMAIN_UNAVAILABLE",
    "The assistant could not load its configuration. Please try again shortly.",
    retryable=True,
    http_status=503,
)

# ── Models ──────────────────────────────────────────────────────────────────

MODEL_UNAVAILABLE = _failure(
    "MODEL_UNAVAILABLE",
    "The AI service is unavailable right now. Please try again in a moment.",
    retryable=True,
    http_status=503,
)
MODEL_TIMEOUT = _failure(
    "MODEL_TIMEOUT",
    "That took too long to answer. Please try again, or ask something narrower.",
    retryable=True,
    http_status=504,
)
RATE_LIMITED = _failure(
    "RATE_LIMITED",
    "Too many requests at once. Please wait a few seconds and try again.",
    retryable=True,
    http_status=429,
)

# ── Retrieval and tools ─────────────────────────────────────────────────────

RETRIEVAL_UNAVAILABLE = _failure(
    "RETRIEVAL_UNAVAILABLE",
    "Your notes could not be searched right now. Please try again shortly.",
    retryable=True,
    http_status=503,
)
NOTES_NOT_INDEXED = _failure(
    "NOTES_NOT_INDEXED",
    "Your notes are still being prepared for search. This usually takes a "
    "moment after you write or edit a note - please try again shortly.",
    retryable=True,
    http_status=200,
)
TOOL_UNAVAILABLE = _failure(
    "TOOL_UNAVAILABLE",
    "A tool the assistant needs is unavailable. Please try again in a moment.",
    retryable=True,
    http_status=503,
)
TOOL_FAILED = _failure(
    "TOOL_FAILED",
    "The assistant could not complete that action. Please try again.",
    retryable=True,
    http_status=502,
)

# ── Deliberate refusals ─────────────────────────────────────────────────────

ACTION_NOT_SUPPORTED = _failure(
    "ACTION_NOT_SUPPORTED",
    "I can't change your notes yet - creating, editing, moving, deleting, and "
    "tagging are still being built. For now I can find things and answer "
    "questions from your notes.",
    retryable=False,
    http_status=200,
)
APPROVAL_REQUIRED = _failure(
    "APPROVAL_REQUIRED",
    "That action needs your confirmation before it can run.",
    retryable=False,
    http_status=200,
)

# ── Anything else ───────────────────────────────────────────────────────────

INTERNAL_ERROR = _failure(
    "INTERNAL_ERROR",
    "Something went wrong on our end. Please try again.",
    retryable=True,
    http_status=500,
)


def get(code: str | None) -> Failure:
    """Look up a failure, falling back to INTERNAL_ERROR for unknown codes.

    Unknown codes are expected during a rolling deploy, when a newer service
    emits a code an older one has not learned yet. Falling back beats raising
    inside an error path.
    """
    return CATALOG.get(code or "", INTERNAL_ERROR)


def classify(exc: BaseException) -> Failure:
    """Map an exception to a failure without importing every service's types.

    Matching is on class name and message rather than on imported classes: this
    module is shared by services that do not install each other's dependencies,
    and an httpx import here would make the catalog un-importable in a service
    that does not use httpx.
    """
    name = type(exc).__name__
    text = str(exc).lower()

    if name in {"ApprovalRequired"}:
        return APPROVAL_REQUIRED
    if name in {"DomainUnavailableError"}:
        return DOMAIN_UNAVAILABLE
    if name in {"McpError", "RerankerUnavailableError"}:
        return TOOL_UNAVAILABLE if "failed:" not in text else TOOL_FAILED
    if name in {"SoftTimeLimitExceeded"} or "timeout" in name.lower() or "timed out" in text:
        return MODEL_TIMEOUT
    if "429" in text or "too many requests" in text or "rate limit" in text:
        return RATE_LIMITED
    if name in {"ConnectError", "ConnectTimeout"} or "connection refused" in text:
        return MODEL_UNAVAILABLE
    if "401" in text or "unauthorized" in text or "403" in text:
        return AUTH_REQUIRED
    return INTERNAL_ERROR


def from_status(status_code: int) -> Failure:
    """Map an upstream HTTP status onto the catalog."""
    if status_code in (401, 403):
        return AUTH_REQUIRED
    if status_code == 422:
        return REQUEST_INVALID
    if status_code == 429:
        return RATE_LIMITED
    if status_code in (502, 503):
        return AGENT_UNAVAILABLE
    if status_code == 504:
        return MODEL_TIMEOUT
    return INTERNAL_ERROR

File: app/frontend/src/lib/failures.js
/**
 * Failure handling for the chat path.
 *
 * The server owns the wording. Every service in the chain (agent-runtime,
 * orchestrator, backend) now sends `{code, message, retryable}` from one
 * catalog in `app/shared/failures.py`, so this module does not keep a second
 * copy of the copy - it renders what it is given and switches behaviour on
 * `code`.
 *
 * The fallbacks below exist only for failures that never reach a server: the
 * fetch itself dying, or an old build still sending bare text. Previously this
 * file's job was done by substring-matching raw Python exception strings
 * (`message.includes('401')`, `includes('inference')`), which broke silently
 * whenever anything upstream reworded an error.
 */

// Codes the UI treats specially. Everything else just renders its message.
export const FAILURE_CODES = {
  AUTH_REQUIRED: 'AUTH_REQUIRED',
  RATE_LIMITED: 'RATE_LIMITED',
  APPROVAL_REQUIRED: 'APPROVAL_REQUIRED',
  ACTION_NOT_SUPPORTED: 'ACTION_NOT_SUPPORTED',
  NOTES_NOT_INDEXED: 'NOTES_NOT_INDEXED',
  CANCELLED: 'CANCELLED',
}

// Used only when the request never reached a service that could classify it.
const CLIENT_FALLBACKS = {
  OFFLINE: 'You appear to be offline. Check your connection and try again.',
  UNKNOWN: 'Something went wrong. Please try again.',
}

/**
 * Normalise anything thrown or streamed into `{code, message, retryable}`.
 *
 * Accepts a server failure payload, an Error carrying one, or a bare Error.
 */
export function toFailure(input) {
  const payload = input?.failure ?? input
  if (payload && typeof payload === 'object' && typeof payload.code === 'string') {
    return {
      code: payload.code,
      message: payload.message || CLIENT_FALLBACKS.UNKNOWN,
      retryable: payload.retryable ?? true,
    }
  }

  // No code: the failure happened in the browser, before any service saw it.
  if (typeof navigator !== 'undefined' && navigator.onLine === false) {
    return { code: 'OFFLINE', message: CLIENT_FALLBACKS.OFFLINE, retryable: true }
  }
  const text = input?.message
  return {
    code: 'UNKNOWN',
    // A server-authored sentence is worth showing; a raw stack-ish string is
    // not, and there is no reliable way to tell them apart, so prefer the
    // generic line unless the text reads like a sentence meant for a person.
    message: looksUserFacing(text) ? text : CLIENT_FALLBACKS.UNKNOWN,
    retryable: true,
  }
}

function looksUserFacing(text) {
  if (typeof text !== 'string' || text.length < 12 || text.length > 200) return false
  if (/https?:\/\/|Error:|Traceback|\bat \w+\.\w+|\{|\}/.test(text)) return false
  return /[.!?]$/.test(text.trim())
}

export function isAuthFailure(failure) {
  return failure?.code === FAILURE_CODES.AUTH_REQUIRED
}

export function isRetryable(failure) {
  return failure?.retryable !== false
}
